RFC111 AI/LLM tool policy: significant revision, limiting drastically their use (#14500)
OSGeo/gdal: doc/source/community/ai_tool_policy.rst

effectively replace the human maintainers, architects, and documenters doing
these jobs who provide much of the value GDAL users actually derive. If they
did, LLM contributors would not need to bother to upstream activity to the
project at all.

With the explosion of LLM usage in software development, the constrained
resource is now "maintenance". It is the time to review your contribution, the time to
make it concise, the time refactor it in to a larger system, and the time to protect
the larger software system from disruption, breakage, and performance
degradation. Indiscriminate usage of LLM in open source projects *consume*
maintenance, and the GDAL LLM tool policy attempts to conserve that
resource.

Policy
--------------------------------------------------------------------------------

Contributors can make **limited use** of LLMs for contributions in GDAL,
subject to details mentioned below:

    * **Human contributors must be the primary author(s) of GDAL contributions**

    * All contributions including code, ticket comments, and commit messages
      should be fully understood by the author(s) submitting them to the
      project.

>     * Submission of unsupervised vibe-coded contributions is *banned*.

Author: What does that mean to "supervise" vibe coded  code ? Just reading the output of the LLM and saying yes "it looks good" ?  That's exactly what we (I) don't want because that doesn't work. And the submitter cannot be the "author" if the code has been written by the LLM

Author: I've added a commit that re-introduces the principle that vibe coding is not allowed + legal aspect

LLM tool policy
================================================================================

Rationale
--------------------------------------------------------------------------------

LLMs are changing software development in many ways. What was once scarce, the
time and expertise to *write* software, is now plentiful. This scarcity was a
choke point that limited consumption of related resources in open source
projects. With that scarcity effectively gone in 2026, the dynamics and economics
of a project like GDAL are being disrupted.

A project like GDAL is not simply code. It is also design, review, coordinated
refactoring, deprecation scheduling, coordinated communication, distribution,
and historical outlook. The plentiful convenience of code writing LLMs do not
effectively replace the human maintainers, architects, and documenters doing
the jobs that provide much of the value GDAL users actually derive. If they
did, LLM contributors would not need to bother to upstream activity to the
project at all.

With the explosion of LLM usage in software development, the constrained
resource is now "maintenance". It is the time to review your contribution, the

> time to make it concise, the time refactor it in to a larger system, and the

Reviewer: ```suggestion
time to make it concise, the time to refactor it in to a larger system, and the
```

time to protect the larger software system from disruption, breakage, and

> performance degradation. Indiscriminate usage of LLM in open source projects

Reviewer: ```suggestion
performance degradation. Indiscriminate usage of LLMs in open source projects
```

*consume* maintenance, and the GDAL LLM tool policy attempts to conserve that
resource.

Additionally, legal systems across the world (including US and EU) have not
definitely determined whether LLM outputs are derived works of training data or
if LLM-written code can even be copyrighted by a human. This is despite it

> being latently extracted and originated from open source software in the first

Reviewer 2: Is "latently" the right word here? Possibly should be "largely"?

place.

Policy
--------------------------------------------------------------------------------

Contributors can make **limited use** of LLMs for contributions in GDAL,
subject to details mentioned below:

    * **Human contributors must be the primary author(s) of GDAL contributions**

    * All contributions including code, ticket comments, and commit messages
      should be fully understood by the author(s) submitting them to the
      project.

>     * Submission of vibe-coded contributions is *banned*.

Reviewer 3: Might be helpful to have a specific definition of "vibe-coded" in this context

Reviewer 4: "I know it when I see it"

Author: https://en.wikipedia.org/wiki/Vibe_coding

Reviewer 3: Maybe make `vibe-coded` into a link to the wikipedia article?

Author: > Maybe make vibe-coded into a link to the wikipedia article?

done

>     * Submission of vibe-coded contributions is *banned*.

Author: ```suggestion
    * Submission of `vibe-coded <https://en.wikipedia.org/wiki/Vibe_coding>`__ contributions is *banned*.
```

>     * Submission of `vibe-coded <https://en.wikipedia.org/wiki/Vibe_coding>`__ contributions is *banned*.

Reviewer 5: It seems to me that an explicit definition of what consitutes "vibe coding" should be included directly in this policy rather than as a reference to Wikipedia.

There are probably also several degrees of "vibe coding".
A developer might start a work item as "vibe coding" and then spend several hours or days refining the code (with various degrees of AI assistance) to the point that the problems inherent to "vibe coding" may no longer be problems?

I am really concerned that a blanket ban on vibe-coding for open-source projects are problematic, as  closed-source solutions would likely allow whatever achieves best development velocity (still taking maintenance cost into consideration).

Author: > It seems to me that an explicit definition of what consitutes "vibe coding" should be included directly in this policy rather than as a reference to Wikipedia.

We may argue about the exact definition, but fundamentally using LLMs is an anti-social / anti-ethical behavior in the context of open source software were maintainers are still human. That make them even more the contention point that before. As raised by @dbaston https://lists.osgeo.org/pipermail/gdal-dev/2026-May/061636.html we review code from others in the hope that a small percentage of them will grow as maintainers and will take a bit of that burden. Reviewing LLM generated code is just helping tech giants improve their model and doesn't grow any new maintainer.

>  as closed-source solutions would likely allow whatever achieves best development velocity (still taking maintenance cost into consideration).

By that metrics, closed-source solutions have always been "better" than us because they can align more developers. Let them do their thing and we'll talk again about the end result they've achieved in 5 years

Why is velocity so important ? Are there so many missing features in GDAL that they need to be rushed ?
Code added is more a long-term *liability* than an asset.

If velocity of development is so important then people can contribute to Oxigdal, and it is written in Rust. Or create their forked GDAL with agents reviewing & automatically merging code.

By the way, it is great to see so many people reviewing this PR. I'd wish they'd do the same for the more "boring" ones 😜

Reviewer 5: >  fundamentally using LLMs is an anti-social / anti-ethical behavior in the context of open source software where maintainers are still human

I strongly disagree with that statement (and indicated as much in the OSGeo charter member survey about LLM usage), but I fully understand where you're coming from as a key maintainer with an unbelievably large number of things to review.

> Reviewing LLM generated code is just helping tech giants improve their model and doesn't grow any new maintainer.

While this is true, this should also improve the quality of future PRs.

Isn't maintainers productivity also improved by LLMs?

> Why is velocity so important ? 

Long term if a closed-source solution becomes much more capable than an open-source alternative, the recent open-source adoption gain trend might reverse. It's also about being able to do more with one's time and being able to make more contributions (which can still be of a certain quality).

> I'd wish they'd do the same for the more "boring" ones

I think part of the solution to get more maintainers may actually be through LLM usage for performing human-in-the-loop reviews.

I am not against rejecting contributions where there's clearly no human in the loop, but I think "vibe coding" can actually be a very efficient way get started on implementing new features in particular (even if careful self-review / refinements should be essential).

Reviewer: Going back to the original question of definitions, I would argue that by the time we come up with a good definition for vibe coding, it will already be obsolete. We should also appreciate how quickly this landscape is evolving and treat any policies we come up with as living documents.

Author: Adopting those new toys under the pressure of a feeling of urgency would be a terrible idea. That's exactly the narrative that tech giants want to instigate in our minds. My personal opinion regarding e.g. adopting a new programming language has been "let's way 15 years and see if it is still there" (Rust is almost at that point :-))

GDAL has always been a very conservative project in terms of tech adoption (I have to fight each time I want to bump the C++ version!). I guess if it is still there after 28 years, this must be part of the reason of its success.

Reviewer 6: I agree with Jérôme on providing a definition for "vibe coding". Something simple like: "source code generated by an LLM which the committer is not fully able to understand or explain".

Besides that, @jerstlouis a blanked ban on LLM generated code might well be ultimate result of the current legal and regulatory framework. When it is a company like [Oracle doing so](https://openjdk.org/legal/ai) everyone else should take heed.

Author: >  "which the committer is not fully able to understand or explain".

You still miss the point that even if you can explain the code line by line, that may still be problematic, because it often generates more code than needed, or takes a wrong approach. And it is very hard to debunk it.
LLMs make it far too easy at throwing tons of code at the face of maintainers. This breaks the way projects are structured.

> When it is a company like [Oracle doing so](https://openjdk.org/legal/ai) everyone else should take heed.

One of my customers (big company) also explicitly bans me from using LLMs for GDAL contracted development. But that's not the reason for my opposition. My limited exposure to reviewing LLM driven contributions has been a terrible experience and showed me that it would ultimately lead to burn out if becoming the norm.

Reviewer 6: I see. In that case I suggest dropping the term "vibe coding" altogether and focus directly on LLM assist. The wording used by Oracle is a good starting point, making it clear which activities are allowed and those which are not.

Reviewer 5: > has been a terrible experience and showed me that it would ultimately lead to burn out if becoming the norm.

I really believe the solution for this should not be banning LLM assistance.

A possible solution is to have parallel tracks of pull requests for different level of LLM assistance.

Perhaps indicating the amount of human time invested by the contributor(s) in preparing a PR.

Each maintainer could decide which of these tracks they invest how much efforts in.

Author: > Each maintainer could decide which of these tracks they invest how much efforts in.

I haven't heard about any existing GDAL maintainer who was keen in reviewing LLM generated PRs. As a proof of that, notice the 3 ones flagged as such at the bottom of https://github.com/OSGeo/gdal/pulls that have been sitting there for many weeks. It is disingenuous to say to potential contributors that they may use LLM assisted coding if there are no maintainers willing to review such PR.

GDAL is already a much too big beast compared to the size of its maintainer community. We don't need more code coming from people interested in drive-by contributions.

Reviewer 7: There was an initial version of this policy, at it was confirmed that is was not enough. If I remember correctly it was "softened" after some discussions. Now we have to harden it.

It has been explained many times: the reviewing and maintenance resources are very, very limited. The experience is that throwing LLM generated code saturates those resource. So we have to cut this. Immediately. Otherwise the project CANNOT CONTINUE.

GDAL is not in a hurry to implement anything. It is mature, stable and well known (and good quality). And these qualities are valued out there, by open and proprietary code.

For sure in some time we will review this policy. Maybe to make it even harder, or softer. Let's see. But now we have to act to protect the project and its maintainers.

Reviewer 8: https://dl.acm.org/doi/full/10.1145/3807518 might be interesting to this discussion. They offer their own definition, and explain why, based on current literature, vibe coding in large, long-lived projects can be problematic.

>     * LLM use may only be used as an improved auto-completion mechanism, or

Reviewer 3: ```suggestion
    * LLMs may only be used as an improved auto-completion mechanism, or
```

      for repeated tasks (mechanical refactoring) that could potentially be
      completed with a deterministic algorithm.

    * Human-coordinated or uncoordinated (OpenClaw, etc) use of agents for
      submission of contributions to the GDAL repository is *banned*.

    * *Any* LLM usage must be indicated by ticket label, comment, or commit
      message indication and account for what was written by whom/what.

    * The contributing human author is ultimately responsible for every line of
      code, comment, or mailing list interaction they initiate, and all of it

>       are subject to the project's :ref:`code_of_conduct`.

Reviewer: ```suggestion
      is subject to the project's :ref:`code_of_conduct`.
```

Summary
-------

GDAL's policy is that contributors can make **very limited use** of
AI (Artificial Intelligence) / Large Language Model (LLM) for their contributions,
subject to details mentioned below, with the general principle that
**there must be a human in the loop**.

Code contributions
------------------

"Vibe coding", or any code generated by a "prompt" is banned. Use of "agents" is banned.
AI/LLM use may only be used as an improved auto-completion mechanism, or
for repeated tasks (mechanical refactoring) that could be completed with non-LLM tools.
Any use of them must be indicated in the pull request, with the details of what
has been used.
The developer must be the author of each line of code they produce and is fully
accountable for their contributions.

Commit messages and pull request messages must be fully written by the author,

> besides potential translation to English and typo/grammar fixing. This is

Reviewer 4: We'll see how this develops, but I suspect it will encourage contributors to post LLM-(re)written walls of text under the pretense of fixing the grammar.

to limit the verbosity of AI/LLM tool.

Issues, mailing list posts, RFCs, design proposals
---------------------------------------------------

The content must be written by a human. Use of AI/LL tool for translation
to English and typo/grammar fixing is allowed.
If an issue was found or analyzed by a AI/LLM tool, the reporter must summarize
the information with their own word and not post the raw output from the tool.
Reproducer code or data sample may be generated by AI/LLM tool.

Rationale for above policy
--------------------------

Besides all what can be read in the wild questioning LLM use, the following
points are particularly relevant:

* The current state of law (at least in US jurisdiction) is that code generated

>   by LLMs cannot be subject to copyright, as it has not been written by a human.

Reviewer 9: ...by LLMS *is not eligible for copyright protection*, as it...

The point is that not being generated by a human does not magically resolve all derived work issues.  It is merely that the new work is not eligible.   To reduce towards absurdity, if you take a 1 page essay whose words are subject to copyright (a human work, recent), and then photocopy it, you can't claim copyright on the novel aspects of photocopying.  But in order to legally copy the copy, you still need permission from the original copyright holder.

Then please add

It is not a settled question whether and to what extent LLM output is legally a derived work of data in the training set, and thus not settled if one needs a license from the copyright holders of the training data to copy/modify/distribute the output.   While many act as if LLM output is not a derived work, there was recently a [significant settlement in the US](https://authorsguild.org/advocacy/artificial-intelligence/what-authors-need-to-know-about-the-anthropic-settlement/).

Author: I don't want to complicate the debate about the legality.  The current phrasing proposed by @lnicola (not sure why your comment is based on my initial phrasing and not his) is sufficient enough for the purpose of this policy.

Reviewer 9: It's probably based on that because that was in my browser window after dinner, and when I hit + it didn't tell me things had been modified/rebased.

The text you commented "thanks, adopted" is great.   It refers to both points tersely, declining to argue.  It's better than what I proposed.

Summary
-------

GDAL's policy is that contributors can make **very limited use** of
AI (Artificial Intelligence) / Large Language Model (LLM) for their contributions,
subject to details mentioned below, with the general principle that

> **the human must be the (primary) author**.

Reviewer 10: I was wondering if we can put it this way: 
```suggestion
**the human must be the (primary) author** , all the AI generated code is fully understood by the author and could have been written by the author as it is.
```

Code contributions
------------------

"Vibe coding", or any code generated by a "prompt" is banned. Use of "agents" is banned.
AI/LLM use may only be used as an improved auto-completion mechanism, or

> for repeated tasks (mechanical refactoring) that could be completed with non-LLM tools.

Reviewer 10: ```
... that could be completed with non-LLM tools.
```
isn't this true for every contribution? What do you mean exactly?

Author: > What do you mean exactly?

a prompt like "switch the second and third argument of the call to foo()" that could be done with some regex or possibly some advanced tool of a C++/Python syntax aware editor is fine

Any use of them must be indicated in the pull request, with the details of what
has been used.
The developer must be the author of each line of code they produce and is fully
accountable for their contributions.

Commit messages and pull request messages must be fully written by the author,
besides potential translation to English and typo/grammar fixing. This is
to limit the verbosity of AI/LLM tool.

Issues, mailing list posts, RFCs, design proposals
---------------------------------------------------

The content must be written by a human. Use of AI/LLM tool for translation
to English and typo/grammar fixing is allowed.
If an issue was found or analyzed by a AI/LLM tool, the reporter must summarize
the information with their own word and not post the raw output from the tool.

> Reproducer code or data sample may be generated by AI/LLM tool.

Reviewer 10: Does this refer to automated tests? If that's the case I would probably move it to the general section and create a sub-paragraph specifically for the testing code.

Author: > Does this refer to automated tests?

no, it refers to someone creating an issue and providing a code/data snippet to reproduce it.
We don't want regression test code to be vibe coded because it could become bloated.

to English and typo/grammar fixing is allowed.
If an issue was found or analyzed by a AI/LLM tool, the reporter must
confirm the findings of those tools with an actual reproducing test case,
not just a theoretical analysis, and summarize the information with their own
words and not post the raw output from the tool.
Reproducer code or data sample may be generated by AI/LLM tool.

Rationale for above policy
--------------------------

Besides all what can be read in the wild questioning LLM use, the following
points are particularly relevant:

* Courts (in particular the US ones) have not definitely determined whether
  LLM outputs are derived works of the training data, or whether LLM-written
  code can even be copyrighted by a human.

* Vibe coding inverts the traditional amount of effort in the contribution process by
  requiring more effort from the reviewer than the contributor. This is neither
  sustainable nor ethical (it makes human reviewers slaves of machines).

* Vibe coding generally generates more code than what a human could come with.
  This tends to produce bloated code bases, which make them unmaintainable in

>   the long run, without forcing other contributor to use AI/LLM tools.

Reviewer: ```suggestion
  the long run, without forcing other contributors to use AI/LLM tools.
```

    AI-generated contributions:
    https://gdal.org/community/ai_tool_policy.html

If a contributor fails to rectify their contribution to comply with the policy,
maintainers may lock the conversation and/or close the pull request/issue/RFC.
In case of repeated violations of our policy, the GDAL project reserves itself
the right to temporarily or permanently ban the infringing person/account.

Mitigation
--------------------------------------------------------------------------------

The GDAL :ref:`sponsorship_program` is one way your organization can help
buffer the cost and disruption of LLMs in keystone projects such as GDAL. The
constrained resource is maintenance, not adding more code/capability. The
Sponsorship Program financially supports operation of GDAL as an ongoing open source
software project, and without it, much of the activity GDAL users take for granted
would simply not happen.

.. below is an allow-list for spelling checker.

.. spelling:word-list::
    LLM
    documenters

>     unmerged

Author: ```suggestion
    unmerged
    latently
```
